clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -221,21 +311,76 @@ private boolean flushNetOutBuffer() throws IOException {
         return netOutBuffer.completed();
     }
 
-    public static class ClientCallbackHandler implements CallbackHandler {
+    private void handleKafkaResponse(RequestHeader requestHeader, byte[] responseBytes) {
+        ByteBuffer responseBuffer = ByteBuffer.wrap(responseBytes);
+        ResponseHeader responseHeader = ResponseHeader.parse(responseBuffer);
+        if (responseHeader.correlationId() != requestHeader.correlationId()) {
+            throw new IllegalStateException("Correlation id for response (" + responseHeader.correlationId()
+                    + ") does not match request (" + requestHeader.correlationId() + ")");
+        }
+        Struct struct;
+        try {
+            struct = ProtoUtils.responseSchema(requestHeader.apiKey(), requestHeader.apiVersion()).read(responseBuffer);
+        } catch (SchemaException e) {
+            LOG.debug("Invalid SASL mechanism response, server may be expecting only GSSAPI tokens");
+            throw new AuthenticationException("Invalid SASL mechanism response", e);
+        }
+        ApiKeys apiKey = ApiKeys.forId(requestHeader.apiKey());
+        switch (apiKey) {
+            case SASL_HANDSHAKE:
+                handleSaslHandshakeResponse(new SaslHandshakeResponse(struct));
+                break;
+            default:
+                throw new IllegalStateException("Unexpected API key during handshake: " + apiKey);
+        }
+    }
+
+    private void handleSaslHandshakeResponse(SaslHandshakeResponse response) {
+        if (response.errorCode() == Errors.UNSUPPORTED_SASL_MECHANISM.code()) {
+            throw new UnsupportedSaslMechanismException(String.format("Client saslMechanism '%s' not enabled in the server, enabled mechanisms are %s",
+                    mechanism, response.enabledMechanisms()));
+        } else if (response.errorCode() != 0) {
+            throw new AuthenticationException(String.format("Unknown error code %d, client mechanism is %s, enabled mechanisms are %s",
+                    response.errorCode(), mechanism, response.enabledMechanisms()));
+        }
+    }
+
+    public static class ClientCallbackHandler implements AuthCallbackHandler {
```

Comment:
Renamed and moved to top-level class, consistent with SaslServerCallbackHandler.